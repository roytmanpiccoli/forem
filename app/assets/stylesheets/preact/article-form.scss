@import '../variables';
@import '../mixins';

.article-form-video-preview {
  text-align: center;
  padding-top: 60px;
  margin: auto;
  max-width: 880px;
  .article-form-video-image-url {
    padding-top: 15px;
    font-family: var(--ff-monospace);
    input {
      padding: 3px;
      font-size: 1em;
      border-radius: 3px;
      border: 1px solid darken($light-medium-gray, 5%);
      border: var(
        --theme-container-border,
        1px solid darken($light-medium-gray, 5%)
      );
      width: 340px;
      margin: 10px;
      &[type='submit'] {
        padding: 3px 8px;
        border: 1px solid $green;
        border-radius: 100px;
        width: 200px;
      }
    }
  }
}

.articleform__uploaderror {
  color: darken($red, 8%);
  font-size: 0.8em;
}
