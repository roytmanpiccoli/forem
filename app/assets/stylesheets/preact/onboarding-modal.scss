@import '../variables';
@import '../mixins';
@import 'components/buttons';
@import 'components/forms';
@import 'config/import';

#onboarding-container {
  min-height: calc(100vh - 350px);
}

.pages-onboarding {
  min-height: initial;
}

// modal box card
.onboarding-main {
  border: none;

  @media screen and (min-width: 600px) {
    border-width: 2px;
  }

  a {
    color: var(--link-branded-color);
  }
}

// modal inner content spacing
.onboarding-content {
  box-sizing: border-box;
  padding: var(--su-4);
  width: 100%;
  min-height: 100%;
  overflow: auto;

  @media screen and (min-width: $breakpoint-s) {
    justify-content: center;
    padding: var(--su-7) var(--su-8);
  }
}

// terms and conditions view
.terms-and-conditions-wrapper {
  $content-height: 358px;
  height: auto;

  button {
    @extend .crayons-btn;
  }
  .terms-and-conditions-content {
    height: $content-height;
    overflow-y: scroll;
    margin-top: 10px;
  }
}

// window background
.onboarding-body {
  -moz-background-size: cover;
  -o-background-size: cover;
  -webkit-background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  display: flex;
  height: 100vh;
  justify-content: start;
  line-height: 200%;
  margin: 0;
  overflow-y: scroll;
  position: fixed;
  top: 0;
  width: 100%;

  @media screen and (min-width: 600px) {
    justify-content: center;
    align-items: center;
    overflow: hidden;
  }

  &:before {
    display: block;
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: rgba(0, 0, 0, 0.65);
    height: 100%;
    width: 100%;
  }
}

// modal navigation
.onboarding-navigation {
  align-self: flex-end;
  flex-shrink: 0;
  width: 100%;
  border-bottom: 2px solid #000;

  &.intro-slide {
    border-bottom: none;
  }

  .navigation-content {
    display: flex;
    justify-content: space-between;
    padding: var(--su-4);

    &.intro-slide {
      padding: 0;
    }

    button {
      @extend .crayons-btn;
    }

    .back-button-container,
    .next-button {
      // Explicitly set width to avoid next-button re-sizing with text,
      // and to ensure that back-button-container takes up the same amount of space.
      width: var(--su-10);
    }

    .back-button {
      @extend .crayons-btn;
      @extend .crayons-btn--ghost;
      @extend .crayons-btn--icon;

      $button-size: 40px; // Explicitly set size for accessibility.

      border-radius: 100%;
      min-width: $button-size;
      height: $button-size;
      align-items: center;
      display: flex;

      &:hover,
      &:active {
        background: var(--base-10);
      }
    }

    .skip-for-now {
      @extend .crayons-btn;
      @extend .crayons-btn--secondary;
    }

    .stepper {
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .dot {
      $dot-size: 10px;

      background-color: var(--base-30);
      border-radius: 50%;
      width: $dot-size;
      height: $dot-size;
      margin: var(--su-1);
    }

    .active {
      background-color: var(--accent-brand);
    }
  }
}

// modal header
.onboarding-content-header {
  margin-bottom: var(--su-4);

  @media screen and (min-width: $breakpoint-s) {
    margin-bottom: var(--su-6);
  }

  .title,
  .subtitle {
    margin: 0;
  }

  .title {
    font-weight: var(--fw-heavy);
    margin-bottom: var(--su-1);
    text-align: left;
  }

  .subtitle {
    color: var(--card-color-secondary);
    font-size: var(--fs-xl);
    font-weight: var(--fw-medium);
    letter-spacing: normal;
    line-height: var(--lh-base);
  }
}

// override crayons notice
.onboarding-body {
  .crayons-notice {
    display: block;
    margin-bottom: var(--su-2);
  }
}
// modal overflow scrolling
.onboarding-modal-scroll-container {
  height: calc(100vh - 289px);
  overflow-y: scroll;
  border-bottom: 1px solid var(--base-20);

  @media screen and (min-height: 800px) {
    max-height: 520px;
  }
}

// tag styles
.onboarding-tags {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: var(--su-4);
  width: 100%;

  @media screen and (min-width: 800px) {
    grid-template-columns: repeat(3, 1fr);
  }

  &__button {
    @extend .crayons-btn;
    border: 0;
    display: block;
    position: absolute;
    left: var(--su-4);
    right: var(--su-4);
    bottom: var(--su-4);
    width: calc(100% - 32px);
    transform: translateY(0px);
    font-family: var(--ff-sans-serif);
    $button: &;

    @media screen and (min-width: 800px) {
      transform: translateY(60px);
    }

    &-default {
      display: inline-block;
    }

    &-alt {
      display: none;
    }
  }

  &__item {
    border-radius: var(--radius);
    position: relative;
    overflow: hidden;
    height: 150px;
    $item: &;

    &__inner {
      background: var(--base-10);
      bottom: 0;
      font-family: var(--ff-monospace);
      hyphens: auto;
      left: 0;
      padding: var(--su-4);
      position: absolute;
      right: 0;
      top: 0;
      transition: all ease-in-out 0.1s;
    }

    &:hover,
    &:focus-within {
      #{$item}__inner {
        background: transparent;
      }

      .onboarding-tags__button {
        transform: translateY(0);

        &:hover,
        &:focus-within {
          transform: translateY(-2px);
        }
      }
    }

    &--selected {
      #{$item}__inner {
        background: transparent;
      }

      input[type='checkbox']:checked {
        background: var(--link-branded-color);
      }
    }
  }
}

// Onboarding Intro
.onboarding-main.introduction {
  .onboarding-content {
    background: var(--base-100);
    color: white;
    display: flex;
    flex-direction: column;
    justify-content: start;
    padding: var(--su-8);
    height: auto;

    @media screen and (min-width: $breakpoint-s) {
      justify-content: center;
    }
  }

  figure {
    margin: 0 0 var(--su-2) 0;

    .sticker-logo {
      height: var(--su-8);
      width: var(--su-8);
      box-sizing: content-box;
      transform: rotate(-30deg);
      padding: var(--su-3);
      vertical-align: middle;
    }
  }

  .introduction-title,
  .introduction-subtitle {
    margin: 0;
  }

  .introduction-title {
    font-size: var(--fs-2xl);
    font-weight: var(--fw-heavy);
    line-height: var(--lh-tight);
    margin-bottom: var(--su-1);

    @media (min-width: $breakpoint-s) {
      font-size: var(--fs-4xl);
    }
  }

  .introduction-subtitle {
    font-size: var(--fs-l);
    font-weight: var(--fw-normal);
    line-height: var(--lh-base);

    @media (min-width: $breakpoint-s) {
      font-size: var(--fs-xl);
    }
  }

  .navigation-content {
    padding: var(--su-7);
    justify-content: center;

    &.intro-slide {
      padding: 0;
    }

    button {
      @extend .crayons-btn;
    }
  }
}

// follow users slide
$onboarding-user-unselected-hover: rgba(71, 85, 235, 0.05);
$onboarding-user-selected: rgba(71, 85, 235, 0.1);
$onboarding-user-selected-hover: rgba(71, 85, 235, 0.2);

.onboarding-body {
  .content-row {
    background-color: var(--base-inverted); // Specify background for Safari.
    align-items: center;
    border: none;
    box-sizing: border-box;
    color: var(--body-color);
    display: flex;
    padding: var(--su-2) var(--su-4);
    text-align: left;
    width: 100%;

    &.unselected {
      &:hover {
        background-color: $onboarding-user-unselected-hover;
        cursor: pointer;
      }

      button {
        @extend .crayons-btn;
        @extend .crayons-btn--outlined;
      }
    }

    &.selected {
      background-color: $onboarding-user-selected;

      &:hover {
        background-color: $onboarding-user-selected-hover;
        cursor: pointer;
      }

      button {
        @extend .crayons-btn;
      }
    }
  }

  .user-avatar-container {
    align-self: baseline;
    margin: 0 var(--su-4) 0 0;

    @media screen and (min-width: $breakpoint-s) {
      align-self: center;
    }
  }

  .user-avatar {
    border-radius: 100%;
    height: var(--su-8);
    width: var(--su-8);
    object-fit: cover;
    border: 1px solid var(--base-10);
  }

  .current-user-info {
    text-align: center;
    margin-bottom: 1rem;

    .current-user-avatar-container {
      height: 80px;
    }

    .current-user-avatar {
      border-radius: 100%;
      height: inherit;
      object-fit: cover;
      border: 2px solid var(--base-90);
    }

    h3 {
      margin-top: var(--su-2);
      line-height: var(--lh-tight);
    }

    p {
      font-size: var(--fs-base);
      line-height: var(--lh-tight);
      color: var(--base-70);
    }
  }

  .user-name,
  .user-summary {
    margin: 0;
    line-height: var(--lh-tight);
    overflow-wrap: anywhere;
  }

  .user-name {
    font-size: var(--fs-base);
  }

  .user-info {
    width: 100%;
  }

  .user-summary {
    font-size: var(--fs-s);
  }

  .user-following-status {
    margin-left: var(--su-7);
  }
}

// closing slide
.onboarding-what-next {
  display: flex;
  flex-direction: column;
  overflow: auto;
  min-height: 250px;

  @media screen and (min-width: $breakpoint-s) {
    flex-direction: column;
  }

  p {
    align-items: center;
    display: inline-flex;
  }

  a {
    background: #d9e7ff;
    margin: 4px auto;
    border-radius: 3px;
    display: block;
    font-weight: 800;
    width: 100%;
    padding: 0.25em;
    position: relative;
    text-align: center;
    box-sizing: border-box;

    .whatnext-emoji {
      text-align: center;
      font-size: 32px;
      margin-right: 8px;
    }
  }
}

.onboarding-previous-location {
  text-align: center;
  margin-top: 10px;
  font-weight: bold;
  display: none;

  @media screen and (min-width: 600px) {
    display: block;
  }

  code {
    font-size: 0.6em;
    margin-top: -10px;
    display: block;
    font-weight: 400;
  }
}

// Follow user
.onboarding-body {
  .content-row {
    align-items: center;
    color: #202428;
    display: flex;
    padding: 8px 16px;
    text-align: left;
    width: 100%;

    &:hover,
    &:focus-within {
      background-color: #eef0f1 !important;
    }
  }

  .user-avatar-container {
    margin: 0 16px 0 0;
  }

  .user-avatar {
    border-radius: 100%;
    height: 48px;
    width: 48px;
    object-fit: cover;
    border: 1px solid #eef0f1;
  }

  .user-name,
  .user-summary {
    margin: 0;
  }
}

// Forms
.onboarding-main {
  fieldset {
    padding: 0;
    border: none;
    margin-bottom: var(--su-6);

    &:last-child {
      margin-bottom: 0;
    }

    ul {
      list-style-type: none;
      margin: 0;
      padding-left: 0;
    }
  }

  legend {
    color: var(--body-color);
    font-weight: var(--fw-medium);
  }

  input[type='text'],
  textarea {
    @extend .crayons-textfield;
    box-sizing: border-box;
  }

  textarea {
    min-height: 77px;
  }

  label {
    @extend .crayons-field;
    @extend .crayons-field__label;

    a {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  .checkbox-form-wrapper {
    padding: var(--su-7);

    .checkbox-form {
      padding-bottom: var(--su-6);
    }
  }

  .crayons-field--checkbox {
    align-items: center;
  }

  // Checkboxes
  .checkbox-item {
    label {
      @extend .crayons-field__label;
      align-items: baseline;
      border-radius: var(--radius);
      display: flex;
      flex-direction: row;
      padding: 0 var(--su-1);
      width: 100%;

      @media screen and (min-width: $breakpoint-s) {
        align-items: center;
      }

      &:hover {
        background: var(--body-bg);
        cursor: pointer;
      }

      &:last-child {
        margin-bottom: 0;
      }

      input[type='checkbox'] {
        @extend .crayons-checkbox;
        margin-right: var(--su-2) !important;
      }
    }
  }

  .onboarding-profile-sub-section {
    margin-bottom: var(--su-6);
  }

  @media screen and (max-height: 790px) {
    .onboarding-form-input--last {
      margin-bottom: 25px;
    }
  }
}

// Adjusting modals... This is naughty... Santa won't come.
.onboarding-main {
  --onboarding-modal-height: 800px;

  &.introduction {
    --onboarding-modal-height: unset;
  }

  .crayons-modal__box {
    @media (min-width: $breakpoint-s) {
      height: var(--onboarding-modal-height);
    }
  }
}
