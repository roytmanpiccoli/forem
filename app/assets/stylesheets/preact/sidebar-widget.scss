@import '../variables';
@import '../mixins';

.sidebar-additional {
  .widget {
    .widget-suggested-follows-container {
      h4 {
        margin: 0;
      }
      .widget-body {
        .widget-list-item__suggestions {
          display: flex;
          align-items: center;
          margin-bottom: 8px;

          & + .widget-list-item__suggestions {
            @include themeable(
              border-top,
              theme-border,
              1px solid $light-medium-gray
            );
            padding-top: 8px;
          }

          .widget-list-item__avatar {
            width: 48px;
            height: 48px;

            img {
              width: 100%;
              height: 100%;
              border-radius: 100%;
            }
          }

          .widget-list-item__content {
            flex: 1 auto;
            padding-left: 8px;
            h5 {
              all: unset;
              display: block;
              margin-bottom: 4px;

              a {
                @include themeable(color, theme-color, rgba($black, 0.8));

                &:hover {
                  color: var(--body-color);
                }
              }
            }

            img {
              width: 48px;
              height: 48px;
              border-radius: 50%;
              position: absolute;
              top: unset;
              left: 15px;
            }
            .widget-list-item__follow-button {
              border: 2px solid transparent;
              border-radius: 3px;
              padding: 2px 8px;
              text-align: center;
              background-color: $green;
              color: $black;
              font-family: $helvetica-condensed;
              font-weight: 900;
              font-size: 0.8em;
            }
          }
        }
      }
    }
    .widget-image {
      margin: auto -16px; // offset the padding of the widget for img full width.
    }

    .widget-button {
      display: block;
      text-align: center;
      padding: 8px 16px;
      margin: auto -16px;
      border-radius: 3px;
    }

    .write {
      margin: 8px -16px;
      background: var(--button-primary-bg);
      color: var(--button-primary-color);
    }

    .all-stories {
      @include themeable(border, theme-border, 2px solid $outline-color);
      color: var(--body-color);
      background: var(--base-inverted);
    }
  }
}
