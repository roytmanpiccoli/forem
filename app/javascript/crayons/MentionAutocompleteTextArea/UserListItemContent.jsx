import { h, Fragment } from 'preact';
import PropTypes from 'prop-types';

/**
 * Component used to display user details in the MentionAutocomplete dropdown
 *
 * @param {object} props
 * @param {object} props.user The user data to populate the list item content with
 */
export const UserListItemContent = ({ user }) => (
  <Fragment>
    <span className="crayons-avatar crayons-avatar--l mr-2 shrink-0">
      <img
        src={user.profile_image_90}
        alt=""
        className="crayons-avatar__image "
      />
    </span>

    <div>
      <p className="crayons-autocomplete__name">{user.name}</p>
      <p className="crayons-autocomplete__username">{`@${user.username}`}</p>
    </div>
  </Fragment>
);

UserListItemContent.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.string,
    username: PropTypes.string,
    profile_image_90: PropTypes.string,
  }).isRequired,
};
