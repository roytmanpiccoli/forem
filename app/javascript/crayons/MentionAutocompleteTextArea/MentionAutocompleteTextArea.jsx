import { h, Fragment } from 'preact';
import { useState, useEffect, useRef, useLayoutEffect } from 'preact/hooks';
import { forwardRef } from 'preact/compat';
import PropTypes from 'prop-types';
import {
  Combobox,
  ComboboxInput,
  ComboboxPopover,
  ComboboxList,
  ComboboxOption,
} from '@reach/combobox';
import { UserListItemContent } from './UserListItemContent';
import {
  getMentionWordData,
  getCursorXY,
  useTextAreaAutoResize,
} from '@utilities/textAreaUtils';
import { useMediaQuery, BREAKPOINTS } from '@components/useMediaQuery';

const MIN_SEARCH_CHARACTERS = 2;
const MAX_RESULTS_DISPLAYED = 6;
// Used to ensure dropdown appears just below search text
const DROPDOWN_VERTICAL_OFFSET = '1.5rem';

/**
 * Helper function to merge any additional ref passed to the MentionAutocompleteTextArea with the inputRefs used internally by this component.
 *
 * @param {Array} refs Array of all references
 */
const mergeInputRefs = (refs) => (value) => {
  refs.forEach((ref) => {
    if (ref) {
      ref.current = value;
    }
  });
};

/**
 * Helper function to copy all styles and attributes from the original textarea to new autocomplete textareas before removing the original node
 *
 * @param {object} options
 * @param {element} options.originalNodeToReplace The DOM element that should be replaced
 * @param {element} options.plainTextArea The DOM element to be used in the 'non-autosuggest' state. It will receive all attributes and styles of the original node.
 * @param {element} options.comboboxTextArea The DOM element to be used in the 'autosuggest' state. It will receive all attributes and styles of the original node.
 */
const replaceTextArea = ({
  originalNodeToReplace,
  plainTextArea,
  comboboxTextArea,
}) => {
  const newNodes = [plainTextArea, comboboxTextArea];

  const { attributes } = originalNodeToReplace;
  const { cssText } = document.defaultView.getComputedStyle(
    originalNodeToReplace,
    '',
  );

  newNodes.forEach((node) => {
    // Make sure all attributes are copied to the autocomplete & plain textareas
    Object.keys(attributes).forEach((attributeKey) => {
      node.setAttribute(
        attributes[attributeKey].name,
        attributes[attributeKey].value,
      );
    });

    // Make sure all styles are copied to the autocomplete & plain textareas
    node.style.cssText = cssText;
    // Make sure no transition replays when the new textareas are mounted
    node.style.transition = 'none';
    // Copy any initial value
    node.value = originalNodeToReplace.value;
  });

  // We need to manually remove the original element, as Preact's diffing algorithm won't replace it in render
  originalNodeToReplace.remove();
};

/**
 * A component for dynamically searching for users and displaying results in a dropdown.
 * This component will optionally replace the textarea passed in props, copying all styles and attributes, and allowing for progressive enhancement.
 * The component functions by switching between a normal textarea and a combobox textarea. Both textareas receive the attributes and styles of the replaceElement prop, and only one is presented at a time.
 * A ref may be given to the component, which will be forwarded to the new textarea
 *
 * @param {object} props
 * @param {element} props.replaceElement The textarea DOM element that should be replaced
 * @param {function} props.fetchSuggestions The async call to use for the search
 * @param {object} props.inputProps Any additional props to be attached to the textarea element
 *
 * @example
 * <MentionAutocompleteCombobox
 *    replaceElement={textAreaRef.current}
 *    fetchSuggestions={fetchUsersByUsername}
 * />
 */

export const MentionAutocompleteTextArea = forwardRef(
  (
    { replaceElement, fetchSuggestions, autoResize = false, ...inputProps },
    forwardedRef,
  ) => {
    const [textContent, setTextContent] = useState(
      inputProps.defaultValue ? inputProps.defaultValue : '',
    );
    const [searchTerm, setSearchTerm] = useState('');
    const [cachedSearches, setCachedSearches] = useState({});
    const [dropdownPositionPoints, setDropdownPositionPoints] = useState({
      x: 0,
      y: 0,
    });
    const [selectionInsertIndex, setSelectionInsertIndex] = useState(0);
    const [users, setUsers] = useState([]);
    const [cursorPosition, setCursorPosition] = useState(null);
    const [ariaHelperText, setAriaHelperText] = useState('');
    const [focusable, setFocusable] = useState(false);

    const isSmallScreen = useMediaQuery(`(max-width: ${BREAKPOINTS.Small}px)`);

    const plainTextAreaRef = useRef(null);
    const comboboxRef = useRef(null);
    const popoverRef = useRef(null);
    const containerRef = useRef(null);

    const { setTextArea, setAdditionalElements } = useTextAreaAutoResize();

    const {
      onChange,
      onBlur,
      id: inputId,
      ...autocompleteInputProps
    } = inputProps;

    useLayoutEffect(() => {
      if (autoResize && comboboxRef.current && plainTextAreaRef.current) {
        setTextArea(plainTextAreaRef.current);
        setAdditionalElements([comboboxRef.current, containerRef.current]);
      }
    }, [autoResize, setTextArea, setAdditionalElements]);

    useEffect(() => {
      if (searchTerm.length < MIN_SEARCH_CHARACTERS) {
        return;
      }

      if (cachedSearches[searchTerm]) {
        setUsers(cachedSearches[searchTerm]);
        return;
      }

      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
        // If the fetchSuggestion call yields more than the MAX, truncate the results
        const resultLength = Math.min(
          fetchedUsers.length,
          MAX_RESULTS_DISPLAYED,
        );

        const results = fetchedUsers.slice(0, resultLength);

        setCachedSearches({
          ...cachedSearches,
          [searchTerm]: results,
        });

        setUsers(results);

        // Let screen reader users know a list has populated
        if (!ariaHelperText && fetchedUsers.length > 0) {
          setAriaHelperText(
            `Mention user, ${fetchedUsers.length} results found`,
          );
        }
      });
    }, [searchTerm, fetchSuggestions, cachedSearches, ariaHelperText]);

    useLayoutEffect(() => {
      const popover = popoverRef.current;
      if (!popover) {
        return;
      }

      const closeOnClickOutsideListener = (event) => {
        if (!popover.contains(event.target)) {
          // User clicked outside, reset to not searching state
          setSearchTerm('');
          setAriaHelperText('');
          setUsers([]);

          const { selectionStart } = comboboxRef.current;

          // Switch back to the plain text area
          comboboxRef.current.classList.add('hidden');
          plainTextAreaRef.current.classList.remove('hidden');
          setCursorPosition(selectionStart + 1);
        }
      };

      document.addEventListener('click', closeOnClickOutsideListener);

      return () =>
        document.removeEventListener('click', closeOnClickOutsideListener);
    }, [searchTerm]);

    useLayoutEffect(() => {
      const { current: plainTextInput } = plainTextAreaRef;
      const { current: combobox } = comboboxRef;

      const activeInput = combobox.classList.contains('hidden')
        ? plainTextInput
        : combobox;

      if (
        focusable ||
        document.activeElement === combobox ||
        document.activeElement === plainTextInput
      ) {
        // Check if the currently focused element is one of the mention autocomplete's
        // inputs. This check is necessary to prevent an issue in iOS browsers only.
        // An additional check to see if the component can be focusable
        // covers the use case for clicking on elements in the mention autocomplete list.
        activeInput.focus({ preventScroll: true });
        activeInput.setSelectionRange(cursorPosition, cursorPosition - 1);
        setFocusable(true);
      }
    }, [cursorPosition, focusable]);

    const handleTextInputChange = ({ target: { value } }) => {
      setTextContent(value);
      const isComboboxVisible =
        !comboboxRef.current.classList.contains('hidden');
      const currentActiveInput = isComboboxVisible
        ? comboboxRef.current
        : plainTextAreaRef.current;

      const { isUserMention, indexOfMentionStart } =
        getMentionWordData(currentActiveInput);

      const { selectionStart } = currentActiveInput;

      if (isUserMention) {
        // search term begins after the @ character
        const searchTermStartPosition = indexOfMentionStart + 1;

        const mentionText = value.substring(
          searchTermStartPosition,
          selectionStart,
        );

        const { x: cursorX, y } = getCursorXY(
          currentActiveInput,
          indexOfMentionStart,
        );
        const textAreaX = currentActiveInput.offsetLeft;

        // On small screens always show dropdown at start of textarea
        const dropdownX = isSmallScreen ? textAreaX : cursorX;

        setDropdownPositionPoints({ x: dropdownX, y });
        setSearchTerm(mentionText);
        setSelectionInsertIndex(searchTermStartPosition);

        if (!isComboboxVisible) {
          // This is the start of a fresh search, transfer use from plain textarea to combobox
          comboboxRef.current.classList.remove('hidden');
          plainTextAreaRef.current.classList.add('hidden');
          setCursorPosition(selectionStart + 1);
        }
      } else if (searchTerm) {
        // User has moved away from an in-progress @mention - clear current search
        setSearchTerm('');
        setAriaHelperText('');
        setUsers([]);

        if (isComboboxVisible) {
          // Search has ended, transfer use from combobox back to plain textarea
          comboboxRef.current.classList.add('hidden');
          plainTextAreaRef.current.classList.remove('hidden');
          setCursorPosition(selectionStart + 1);
        }
      }
    };

    const handleComboboxBlur = () => {
      // User has left the textarea, exit combobox functionality without refocusing plainTextArea
      comboboxRef.current.classList.add('hidden');
      plainTextAreaRef.current.classList.remove('hidden');
    };

    const handleSelect = (username) => {
      // Construct the new textArea content with selected username inserted
      const textWithSelection = `${textContent.substring(
        0,
        selectionInsertIndex,
      )}${username} ${textContent.substring(
        comboboxRef.current.selectionStart,
      )}`;

      // Clear the current search
      setSearchTerm('');
      setUsers([]);
      setAriaHelperText('');

      // Update the text area value
      setTextContent(textWithSelection);

      // Switch back to the plain text input
      comboboxRef.current.classList.add('hidden');
      plainTextAreaRef.current.classList.remove('hidden');

      // Allow any other attached change event to receive the updated text
      onChange?.(textWithSelection);

      // Update the cursor to directly after the selection (+2 accounts for the @ sign, and adding a space after the username)
      const newCursorPosition = selectionInsertIndex + username.length + 2;
      setCursorPosition(newCursorPosition);
    };

    useLayoutEffect(() => {
      const { current: comboboxTextArea } = comboboxRef;
      const { current: plainTextArea } = plainTextAreaRef;

      if (comboboxTextArea && plainTextArea && replaceElement) {
        replaceTextArea({
          originalNodeToReplace: replaceElement,
          plainTextArea,
          comboboxTextArea,
        });
        plainTextArea.focus({ preventScroll: true });
      }

      // Initialize the new text areas in the "non-autosuggest" state, hiding the combobox until a search begins
      comboboxTextArea.classList.add('hidden');
    }, [replaceElement]);

    return (
      <Fragment>
        <div aria-live="polite" class="screen-reader-only">
          {ariaHelperText}
        </div>

        <Combobox
          ref={containerRef}
          id="combobox-container"
          data-testid="autocomplete-wrapper"
          onSelect={handleSelect}
          className={`crayons-autocomplete${autoResize ? ' h-100' : ''}`}
        >
          <ComboboxInput
            {...autocompleteInputProps}
            data-gramm_editor="false"
            aria-label="Mention user"
            ref={comboboxRef}
            value={textContent}
            data-mention-autocomplete-active="true"
            as="textarea"
            autocomplete={false}
            onChange={(e) => {
              onChange?.(e);
              handleTextInputChange(e);
            }}
            onBlur={(e) => {
              onBlur?.(e);
              handleComboboxBlur();
            }}
          />

          <textarea
            {...autocompleteInputProps}
            data-gramm_editor="false"
            id={inputId}
            data-mention-autocomplete-active="true"
            ref={mergeInputRefs([plainTextAreaRef, forwardedRef])}
            onChange={(e) => {
              onChange?.(e);
              handleTextInputChange(e);
            }}
            value={textContent}
          />

          {searchTerm && (
            <ComboboxPopover
              ref={popoverRef}
              className="crayons-autocomplete__popover absolute"
              id="mention-autocomplete-popover"
              style={{
                top: `calc(${dropdownPositionPoints.y}px + ${DROPDOWN_VERTICAL_OFFSET}`,
                left: `${dropdownPositionPoints.x}px`,
              }}
            >
              {users.length > 0 ? (
                <ComboboxList>
                  {users.map((user) => (
                    <ComboboxOption
                      key={user.username}
                      value={user.username}
                      className="crayons-autocomplete__option flex items-center"
                    >
                      <UserListItemContent user={user} />
                    </ComboboxOption>
                  ))}
                </ComboboxList>
              ) : (
                <span className="crayons-autocomplete__empty">
                  {searchTerm.length >= MIN_SEARCH_CHARACTERS
                    ? 'No results found'
                    : 'Type to search for a user'}
                </span>
              )}
            </ComboboxPopover>
          )}
        </Combobox>
      </Fragment>
    );
  },
);

MentionAutocompleteTextArea.propTypes = {
  replaceElement: PropTypes.node,
  fetchSuggestions: PropTypes.func.isRequired,
  autoResize: PropTypes.bool,
};
