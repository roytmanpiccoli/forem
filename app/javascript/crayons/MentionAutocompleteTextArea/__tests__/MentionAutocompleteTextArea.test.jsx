import { h } from 'preact';
import { render, waitFor } from '@testing-library/preact';
import '@testing-library/jest-dom';
import { MentionAutocompleteTextArea } from '../MentionAutocompleteTextArea';

describe('<MentionAutocompleteTextArea />', () => {
  const textArea = document.createElement('textarea');
  textArea.setAttribute('aria-label', 'test text area');
  textArea.setAttribute('id', 'test-text-area');
  textArea.value = 'some text';

  beforeAll(() => {
    global.window.matchMedia = jest.fn((query) => {
      return {
        matches: false,
        media: query,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    });
  });

  beforeEach(() => {
    document.body.appendChild(textArea);
  });

  it('should render', () => {
    const { container } = render(
      <MentionAutocompleteTextArea
        replaceElement={textArea}
        fetchSuggestions={async () => ({ result: [] })}
      />,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should replace the given textarea with a hidden combobox and a plain textarea', () => {
    const { getByRole } = render(
      <MentionAutocompleteTextArea
        replaceElement={textArea}
        fetchSuggestions={async () => ({ result: [] })}
      />,
    );

    const combobox = getByRole('combobox');

    waitFor(() => expect(combobox).not.toBeVisible());
    expect(combobox.id).toEqual('test-text-area');
    expect(combobox.value).toEqual('some text');
    expect(combobox.getAttribute('aria-label')).toEqual('test text area');

    const textarea = getByRole('textbox');
    expect(textarea).toBeVisible();
    expect(textarea.id).toEqual('test-text-area');
    expect(textarea.value).toEqual('some text');
    expect(textarea.getAttribute('aria-label')).toEqual('test text area');
  });
});
