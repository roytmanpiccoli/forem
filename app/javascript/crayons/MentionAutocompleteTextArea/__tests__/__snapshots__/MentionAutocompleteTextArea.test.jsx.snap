// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MentionAutocompleteTextArea /> should render 1`] = `"<div aria-live=\\"polite\\" class=\\"screen-reader-only\\"></div><div id=\\"combobox-container\\" data-testid=\\"autocomplete-wrapper\\" class=\\"crayons-autocomplete\\" data-reach-combobox=\\"\\" data-state=\\"idle\\"><textarea aria-autocomplete=\\"both\\" aria-controls=\\"listbox--combobox-container\\" aria-expanded=\\"false\\" aria-haspopup=\\"listbox\\" aria-label=\\"test text area\\" role=\\"combobox\\" data-gramm_editor=\\"false\\" data-mention-autocomplete-active=\\"true\\" data-reach-combobox-input=\\"\\" data-state=\\"idle\\" id=\\"test-text-area\\" style=\\"font: -webkit-small-control; text-rendering: auto; letter-spacing: normal; word-spacing: normal; line-height: normal; text-transform: none; text-indent: 0; text-shadow: none; display: inline-block; text-align: start; background-color: white; border: 1px solid; flex-direction: column; resize: auto; cursor: auto; padding: 2px; white-space: pre-wrap; word-wrap: break-word; visibility: visible; transition: none;\\" class=\\"hidden\\"></textarea><textarea data-gramm_editor=\\"false\\" data-mention-autocomplete-active=\\"true\\" aria-label=\\"test text area\\" id=\\"test-text-area\\" style=\\"font: -webkit-small-control; text-rendering: auto; letter-spacing: normal; word-spacing: normal; line-height: normal; text-transform: none; text-indent: 0; text-shadow: none; display: inline-block; text-align: start; background-color: white; border: 1px solid; flex-direction: column; resize: auto; cursor: auto; padding: 2px; white-space: pre-wrap; word-wrap: break-word; visibility: visible; transition: none;\\"></textarea></div>"`;
